kroxylicious-filters/kroxylicious-oauthbearer/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerFilter.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.filter.oauthbearer;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.security.sasl.SaslServer;
+
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.SaslAuthenticateRequestData;
+import org.apache.kafka.common.message.SaslAuthenticateResponseData;
+
+import io.kroxylicious.proxy.filter.FilterContext;
+import io.kroxylicious.proxy.filter.RequestFilterResult;
+import io.kroxylicious.proxy.filter.SaslAuthenticateRequestFilter;
+
+import static org.apache.kafka.common.protocol.Errors.SASL_AUTHENTICATION_FAILED;
+
+public class OauthBearerFilter
+        implements SaslAuthenticateRequestFilter {
+
+    private final SaslServer saslServer;
+
+    public OauthBearerFilter(SaslServer saslServer) {
+        this.saslServer = saslServer;
+    }
+
+    @Override
+    public CompletionStage<RequestFilterResult> onSaslAuthenticateRequest(short apiVersion, RequestHeaderData header,
+                                                                          SaslAuthenticateRequestData request, FilterContext context) {
+        try {
+            saslServer.evaluateResponse(request.authBytes());
+        }
+        catch (Exception e) {
+            SaslAuthenticateResponseData failedResponse = new SaslAuthenticateResponseData()
+                    .setErrorCode(SASL_AUTHENTICATION_FAILED.code())
+                    .setErrorMessage("Token not valid");
+            return context.requestFilterResultBuilder().shortCircuitResponse(failedResponse).withCloseConnection().completed();
```

Comment:
Did you consider a backoff delay?  I think we want to avoid a tight loop between the client and proxy, so delaying the completion by a configurable amount would be something to consider.  You'd need to arrange for the filter to use io.kroxylicious.proxy.filter.FilterFactoryContext#eventLoop to schedule the completion later.

A slightly more advanced version might cache the duff token and use this to drive an progressive backoff that increases the delay imposed on the client submitting the same duff token over and over. You'd need access to the token in the case that the auth fails (something OAuthBearerSaslServer doesn't do).  Maybe clever proxying of the AuthenticateCallbackHandler might be do the trick - I've not fully thought this through.

Reply:
+1 to a backoff.

Reply:
hi @k-wall , @tombentley , to solve this problem, can I introduce a library like `resilience4j` or `Bucket4j` ?

Reply:
In general, we are trying to keep the number of dependencies small.

On resilience4j or bucket4j, I haven't used either of those libraries before, so please tell me how you'd envisage using them?  Which features are important?  When I wrote the comment, the immediate idea that occured to me was it implement a local cache using Caffeine.   Caffeine is already part of the project - we use it in the KMS for Record Encryption.

Separately, I realised after writing the comment that if backoff was too aggressive, the client's going to timeout and retry again anyway.   That'd probably be much harder to diagnose in a production environment than a SASL auth failure.  Very long backoff values might do more harm than good.

Reply:
Understand about the dependency, so I did it with Caffeine only.

So, I implemented a basic implementation that needs rework (and more tests) about the warning you exposed if it's too agressive. I pushed this version to let start a discussion about it if it's something we would like to keep or not.

I agree that, if the limit delay is too high, client will handle a timeout exception as if the proxy is not available, not an SASL Error. But, your remark about introducing a backoff makes sens for me to limit resource overhead on kroxylicious too.

Reply:
I think we might different ideas about the backoff protection.

My idea was that we'd throttle a client who is caught in a loop connecting, presenting a duff token, disconnecting and then starting over again with a **new TCP connection**.  I think the Filter would implement this by keeping a cache of tokens it is seen already and imposing the backoff if it sees the same invalid token more >1.

By the looks of your implementation, you have a different train of thought.  You have scoped your cache to the **connection** (`RateLimitSaslManager#wrap()` creates a new `RateLimitSaslManager` and the Caffeine cache is scoped to the connection).   Are you thinking about [KIP-368](https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate)?   I don't think we delaying responses in the reauth case actually helps. The broker is going to disconnect the client if the reauth fails anyway.  The client can't try to reauth again as the connection will be gone.

Separately, in order to support brokers using KIP-368, the SaslServer object can't be scoped to the connection.  It needs to be created by each SaslHandshake and used/disposed by the SaslAuthenticate. That will support the re-auth use-case.

Reply:
> My idea was that we'd throttle a client who is caught in a loop connecting, presenting a duff token, disconnecting and then starting over again with a new TCP connection. I think the Filter would implement this by keeping a cache of tokens it is seen already and imposing the backoff if it sees the same invalid token more >1.

I totally agree with you, I can rework the implementation to follow this.

> By the looks of your implementation, you have a different train of thought. You have scoped your cache to the connection (RateLimitSaslManager#wrap() creates a new RateLimitSaslManager and the Caffeine cache is scoped to the connection). Are you thinking about [KIP-368](https://cwiki.apache.org/confluence/display/KAFKA/KIP-368%3A+Allow+SASL+Connections+to+Periodically+Re-Authenticate)? I don't think we delaying responses in the reauth case actually helps. The broker is going to disconnect the client if the reauth fails anyway. The client can't try to reauth again as the connection will be gone.

I don't find a way to detect if it's a reauth request or not. Is it important to make a difference between authenticate (first request) and a re-authenticate request ? In any case the subject of this filter is to pre-validate the token no ? If I understand right the KIP-368, in each case server receive an authenticate request.

> Separately, in order to support brokers using KIP-368, the SaslServer object can't be scoped to the connection. It needs to be created by each SaslHandshake and used/disposed by the SaslAuthenticate. That will support the re-auth use-case.

No sure that this behavior is a use case for this filter, as at this time, the SASL connection is not maintained by the proxy, we just would like to relieve the kafka cluster.
In my opinion, this behavior is requested when we would like to handle the SASL connection in place of kafka as in #1206

Reply:
Filters can keep state in member variables.  Kroxylicious guarantees that it is only the Netty event loop thread that invokes filter on*Request/Response methods, so you have the guarantee of thread safety.

How about having `OauthBearerValidationFilter` set a flag once you've seen a successful `SaslAuthenticateResponse`?  It could then special case any subsequent `SaslAuthenticateRequests` (which must be SASL reauths) and let them pass without your validation.  That avoids the need to have to recreate the SaslServer.

Reply:
Okay, I read again the KIP-368 to understand what you mean and what I missed.
I didn't note the queue sequence introduced by the re-authenticate behavior.

I reworked the filter implementation trying to implement what we discussed about.
Moreover, the filter now supports only `OAUTHBEARER` mechanism